src/pipe.js

```diff
@@ -30,5 +34,11 @@ var reverse = require('./reverse');
  *      squareThenDoubleThenTriple(5); //=> 150
  */
 module.exports = function pipe() {
-  return compose.apply(this, reverse(arguments));
+  return reduce(function(f, g) {
+    return curryN(f.length + g.length - 1, function() {
+      var fArgs = take(f.length, arguments);
+      var gArgs = drop(f.length, arguments);
+      return g.apply(this, prepend(f.apply(this, fArgs), gArgs));
+    });
+  }, identity, arguments);
```

Comment:
It was only by expressing composition as a fold that I was able to reveal its inherent elegance (and realize that we needn't disallow composition of zero functions). If we agree that the behaviour is correct I've no problem with someone providing an equivalent imperative implementation.